ndk/src/media/media_codec.rs

```diff
@@ -246,9 +313,133 @@ impl MediaCodec {
+        ) {
+            match panic::catch_unwind(|| {
+                let callback = &mut *(user_data as *mut AsyncNotifyCallback);
+                if let Some(f) = callback.on_output_available.as_mut() {
+                    let buffer_info = BufferInfo {
+                        inner: *buffer_info,
+                    };
+                    f(index as usize, &buffer_info);
+                }
+            }) {
+                Ok(result) => result,
+                Err(_) => abort(),
+            }
+        }
+
+        unsafe extern "C" fn ffi_on_format_changed(
+            _codec: *mut ffi::AMediaCodec,
+            user_data: *mut c_void,
+            format: *mut ffi::AMediaFormat,
+        ) {
+            match panic::catch_unwind(|| {
+                let callback = &mut *(user_data as *mut AsyncNotifyCallback);
+                if let Some(f) = callback.on_format_changed.as_mut() {
+                    let format = MediaFormat {
+                        inner: NonNull::new_unchecked(format),
+                    };
+                    f(&format);
+                }
+            }) {
+                Ok(result) => result,
+                Err(_) => abort(),
+            }
+        }
+
+        unsafe extern "C" fn ffi_on_error(
+            _codec: *mut ffi::AMediaCodec,
+            user_data: *mut c_void,
+            error: ffi::media_status_t,
+            action_code: i32,
+            detail: *const c_char,
+        ) {
+            match panic::catch_unwind(|| {
+                let callback = &mut *(user_data as *mut AsyncNotifyCallback);
+                if let Some(f) = callback.on_error.as_mut() {
+                    f(
+                        MediaError::from_status(error),
+                        ActionCode(action_code),
+                        CStr::from_ptr(detail),
+                    );
+                }
+            }) {
+                Ok(result) => result,
+                Err(_) => abort(),
+            }
+        }
+
+        let (callback, ffi_callback, user_data) = if let Some(callback) = callback {
+            // On Android 12 and earlier, due to faulty null checks, if a callback is not set, but at least one other
+            // callback *is* set, then it will segfault in NdkMediaCodec.cpp when trying to invoke the unset callback.
+            // To work around this we just enable all callbacks and do nothing if the corresponding callback is not set
+            // in AsyncNotifyCallback
+            let ffi_callback = ffi::AMediaCodecOnAsyncNotifyCallback {
+                onAsyncInputAvailable: Some(ffi_on_input_available),
+                onAsyncOutputAvailable: Some(ffi_on_output_available),
+                onAsyncFormatChanged: Some(ffi_on_format_changed),
+                onAsyncError: Some(ffi_on_error),
+            };
+
+            let mut boxed = Box::new(callback);
+            let ptr: *mut AsyncNotifyCallback = &mut *boxed;
```

Comment:
You'd typically want to wrap the `Box` (or other pointer-types) in `Pin` to prevent the allocation from accidentally moving around?